source/NVDAObjects/UIA/__init__.py

```diff
@@ -142,127 +224,203 @@ def _getControlFieldForObject(self, obj):
-			for f in self._getFormatFieldsAndText(tempRange,formatConfig):
-				yield f
-			field=self._getControlFieldForObject(childObj) if childObj else None
-			if field:
-				yield textInfos.FieldCommand("controlStart",field)
-			for x in self._getTextWithFieldsForRange(childObj,childRange,formatConfig):
-				yield x
-			if field:
-				yield textInfos.FieldCommand("controlEnd",None)
-			tempRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_Start,childRange,UIAHandler.TextPatternRangeEndpoint_End)
-		tempRange.moveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,rangeObj,UIAHandler.TextPatternRangeEndpoint_End)
-		for f in self._getFormatFieldsAndText(tempRange,formatConfig):
-			yield f
+			parentFields.append(field)
+		log.debug("Done generating controlFields for parents")
+		log.debug("Yielding control starts for parents")
+		for field in reversed(parentFields):
+			yield textInfos.FieldCommand("controlStart",field)
+		log.debug("Done yielding control starts for parents")
+		del parentElements
+		log.debug("Yielding balanced fields for textRange")
+		# Move through the text range, collecting text and recursing into children
+		tempRange=textRange.clone()
+		tempRange.MoveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,tempRange,UIAHandler.TextPatternRangeEndpoint_Start)
+		if recurseChildren:
+			if log.isEnabledFor(log.DEBUG):
+				log.debug("Child count: %s"%childElements.length)
+				log.debug("Walking children")
+			for index in xrange(childElements.length):
+				childElement=childElements.getElement(index)
+				if not childElement or UIAHandler.handler.clientObject.compareElements(childElement,enclosingElement):
+					log.debug("NULL childElement. Skipping")
+					continue
+				childElement=childElement.buildUpdatedCache(UIAHandler.handler.baseCacheRequest)
+				if log.isEnabledFor(log.DEBUG):
+					log.debug("Fetched child %s (%s)"%(index,childElement.currentLocalizedControlType))
+				childRange=self.obj.UIATextPattern.rangeFromChild(childElement)
+				if not childRange:
+					log.debug("NULL childRange. Skipping")
+					continue
+				if childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_Start,textRange,UIAHandler.TextPatternRangeEndpoint_End)>=0:
+					log.debug("Child at or past end of textRange. Breaking")
+					break
+				origChildRange=childRange.clone()
+				if childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_End,textRange,UIAHandler.TextPatternRangeEndpoint_End)>0:
+					log.debug("textRange ended part way through the child. Crop end of childRange to fit")
+					childRange.MoveEndpointByRange(UIAHandler.TextPatternRangeEndpoint_End,textRange,UIAHandler.TextPatternRangeEndpoint_End)
+				childStartDelta=childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_Start,tempRange,UIAHandler.TextPatternRangeEndpoint_End)
+				if childStartDelta>0 and childRange.CompareEndpoints(UIAHandler.TextPatternRangeEndpoint_Start,textRange,UIAHandler.TextPatternRangeEndpoint_Start)>0:

```

Comment:
I don't quite follow the second part of this check.

1. As I understand it, temp is always >= textRange start when we get here.
2. The first part of this check asks whether child start > temp end.
3. If 2) is true, we have textRange start <= temp end < childStart.
4. Then the second part of the check asks if child start  > textRange start.
5. But given 3), isn't 4) always true, thus making 4) redundant?